### ISSUE TITLE
Agent log crammed with CollectExecutionStatistics(_) error

### ISSUE THREAD

**Issue by carusyte** [chrislusf/gleam#53]
excerpt from the log file:

> 2017/08/22 23:13:40 &{0xc420274000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc420252000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc4201be1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc4201b21c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc420130700}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc4201561c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc4201ba1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc420274000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 in memory 5:1 finished writing f1213273487-d5-s1 0 bytes
> 2017/08/22 23:13:40 in memory 6:0 finished reading f1213273487-d5-s1 0 bytes
> 2017/08/22 23:13:40 &{0xc420388000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 &{0xc42009c540}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 in memory 5:6 finished writing f1213273487-d5-s6 0 bytes
> 2017/08/22 23:13:40 in memory 6:0 finished reading f1213273487-d5-s6 0 bytes
> 2017/08/22 23:13:40 in memory 5:9 finished writing f1213273487-d5-s9 0 bytes
> 2017/08/22 23:13:40 in memory 6:0 finished reading f1213273487-d5-s9 0 bytes
> 2017/08/22 23:13:40 &{0xc4201fe380}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:40 in memory 5:11 finished writing f1213273487-d5-s11 0 bytes
> 2017/08/22 23:13:40 in memory 6:0 finished reading f1213273487-d5-s11 0 bytes
> 2017/08/22 23:13:40 &{0xc420374000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:41 deleting f1213273487-d5-s6
> 2017/08/22 23:13:41 deleting f1213273487-d5-s11
> 2017/08/22 23:13:41 deleting f1213273487-d5-s1
> 2017/08/22 23:13:41 deleting f1213273487-d5-s9
> 2017/08/22 23:13:46 in memory 5:5 starts writing f1750670629-d5-s5 expected reader:1
> 2017/08/22 23:13:46 &{0xc42034c000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 in memory 5:1 starts writing f1750670629-d5-s1 expected reader:1
> 2017/08/22 23:13:46 &{0xc4202a6000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 &{0xc420130380}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 &{0xc420230000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 &{0xc4201c81c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 &{0xc4201ca1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:13:46 in memory 6:0 waits for f1750670629-d5-s1
> 2017/08/22 23:13:46 in memory 6:0 start reading f1750670629-d5-s1
> 2017/08/22 23:13:46 in memory 6:0 waits for f1750670629-d5-s5
> 2017/08/22 23:13:46 in memory 6:0 start reading f1750670629-d5-s5
> 2017/08/22 23:15:15 &{0xc42014e1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:15:15 &{0xc4201c8540}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:15:15 &{0xc420278000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:15:15 &{0xc42023a000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:15:15 in memory 5:5 finished writing f1750670629-d5-s5 0 bytes
> 2017/08/22 23:15:15 in memory 6:0 finished reading f1750670629-d5-s5 0 bytes
> 2017/08/22 23:15:15 in memory 5:1 finished writing f1750670629-d5-s1 0 bytes
> 2017/08/22 23:15:15 in memory 6:0 finished reading f1750670629-d5-s1 0 bytes
> 2017/08/22 23:15:15 &{0xc420380000}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
> 2017/08/22 23:15:15 &{0xc42009c1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable

My application gives me unexpected result running in distributed mode which consists of 4 agent nodes. If I run it with local master and agent, everything seems fine. Not sure if this sort of error is related.
`
2017/08/22 23:15:15 &{0xc42014e1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
`

**Comment by chrislusf** [comment]
network connection problem?

**Comment by carusyte** [comment]
Not likely. When I started the master and those 4 agents(on individual server), the log prints heart beat message and the master's log prints messages about agents connected to it.

**Comment by chrislusf** [comment]
I added some debug message. Please try it and see the output.

**Comment by carusyte** [comment]
now the log reads:

> 2017/08/24 20:36:31 Failed2 to read rima: stat rima: no such file or directory
2017/08/24 20:36:31 Failed to send related files: stat rima: no such file or directory

**Comment by carusyte** [comment]
It seems this failure has something to do with the way I started my program. Since the $GOPATH/bin is in my $PATH environment variable so I just started my program without using the full path.
Now I just modified my launcher shell script and start my program using full path, the log reads something like below:

> 2017/08/24 21:47:03 10.16.51.14:45327 1:0>2017/08/24 21:47:03 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 21:47:03 10.16.51.14:45327 1:0>2017/08/24 21:47:03 executor reportStatus to :45327: executor collect stats to :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 21:47:04 10.16.50.11:45327 2:1-3:1-4:1-5:1> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.11:45327 2:11-3:11-4:11-5:11> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.10:45327 2:9-3:9-4:9-5:9> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.11:45327 2:1-3:1-4:1-5:1>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.11:45327 2:1-3:1-4:1-5:1>2017/08/24 21:47:04 Failed 1 time to start /usr/bin/sh [sh -c /home/ctcs/go/bin/rima -gleam.mapper=m2 -gleam.executor=[::]:35329 -flow.hashcode=4221400531 -flow.stepId=3 -flow.taskId=1 -gleam.profiling=false] []:3:1 Wait error exit status 127.
2017/08/24 21:47:04 10.16.50.11:45327 2:1-3:1-4:1-5:1>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.10:45327 2:9-3:9-4:9-5:9>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.11:45327 2:11-3:11-4:11-5:11>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: No such file or directory
2017/08/24 21:47:04 10.16.50.11:45327 2:11-3:11-4:11-5:11>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: No such file or directory
2017/08/24 21:47:04 10.16.50.10:45327 2:9-3:9-4:9-5:9>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.10:45327 2:4-3:4-4:4-5:4> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.10:45327 2:9-3:9-4:9-5:9>2017/08/24 21:47:04 Failed 1 time to start /usr/bin/sh [sh -c /home/ctcs/go/bin/rima -gleam.mapper=m2 -gleam.executor=[::]:41985 -flow.hashcode=4221400531 -flow.stepId=3 -flow.taskId=9 -gleam.profiling=false] []:3:9 Wait error exit status 127.
2017/08/24 21:47:04 10.16.50.11:45327 2:11-3:11-4:11-5:11>2017/08/24 21:47:04 Failed 1 time to start /usr/bin/sh [sh -c /home/ctcs/go/bin/rima -gleam.mapper=m2 -gleam.executor=[::]:36157 -flow.hashcode=4221400531 -flow.stepId=3 -flow.taskId=11 -gleam.profiling=false] []:3:11 Wait error exit status 127.
2017/08/24 21:47:04 10.16.51.14:45327 2:8-3:8-4:8-5:8> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.10:45327 6:0-7:0> starting with 20 MB memory...
2017/08/24 21:47:04 10.16.50.10:45327 2:4-3:4-4:4-5:4>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.10:45327 2:4-3:4-4:4-5:4>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.50.10:45327 2:4-3:4-4:4-5:4>2017/08/24 21:47:04 Failed 1 time to start /usr/bin/sh [sh -c /home/ctcs/go/bin/rima -gleam.mapper=m2 -gleam.executor=[::]:49163 -flow.hashcode=4221400531 -flow.stepId=3 -flow.taskId=4 -gleam.profiling=false] []:3:4 Wait error exit status 127.
2017/08/24 21:47:04 10.16.51.14:45327 2:8-3:8-4:8-5:8>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.51.14:45327 2:8-3:8-4:8-5:8>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: Nosuch file or directory
2017/08/24 21:47:04 10.16.51.14:45327 2:8-3:8-4:8-5:8>2017/08/24 21:47:04 Failed 1 time to start /usr/bin/sh [sh -c /home/ctcs/go/bin/rima -gleam.mapper=m2 -gleam.executor=[::]:53281 -flow.hashcode=4221400531 -flow.stepId=3 -flow.taskId=8 -gleam.profiling=false] []:3:8 Wait error exit status 127.
2017/08/24 21:47:04 10.16.51.14:45327 2:10-3:10-4:10-5:10> starting with 3 MB memory...
2017/08/24 21:47:04 10.16.50.10:45327 6:0-7:0>/home/ctcs/go/bin/rima: error while loading shared libraries: libclntsh.so.12.1: cannot open shared object file: No such file or directory
2017/08/24 21:47:04 10.16.50.11:45327 2:1-3:1-4:1-5:1>2017/08/24 21:47:04 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 21:47:04 10.16.53.30:45327 2:7-3:7-4:7-5:7> starting with 3 MB memory...

**Comment by carusyte** [comment]
Now I exported the $LD_LIBRARY_PATH in a shell under /usr/etc/profile.d, and restarted gleam, the previous error of unable to load libclntsh.so.12.1 was gone. But the old friend still haunts me:

> 2017/08/24 22:34:14 ##.##.50.11:45327 1:0> starting with 5 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 1:0>2017/08/24 22:34:14 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.50.11:45327 1:0>2017/08/24 22:34:14 executor reportStatus to :45327: executor collect stats to :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.51.11:45327 2:10-3:10-4:10-5:10> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:4-3:4-4:4-5:4> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.51.11:45327 2:8-3:8-4:8-5:8> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:11-3:11-4:11-5:11> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.51.11:45327 2:0-3:0-4:0-5:0> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:2-3:2-4:2-5:2> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:4-3:4-4:4-5:4>2017/08/24 22:34:14 &{0xc4201c81c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.51.11:45327 2:10-3:10-4:10-5:10>2017/08/24 22:34:14 &{0xc4201b4380}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.50.11:45327 2:6-3:6-4:6-5:6> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:11-3:11-4:11-5:11>2017/08/24 22:34:14 &{0xc42013e380}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.51.11:45327 2:8-3:8-4:8-5:8>2017/08/24 22:34:14 &{0xc4201d2380}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.50.11:45327 2:2-3:2-4:2-5:2>2017/08/24 22:34:14 &{0xc4201bc1c0}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.50.10:45327 2:5-3:5-4:5-5:5> starting with 3 MB memory...
2017/08/24 22:34:14 ##.##.50.11:45327 2:4-3:4-4:4-5:4>2017/08/24 22:34:14 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.51.11:45327 2:10-3:10-4:10-5:10>2017/08/24 22:34:14 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable
2017/08/24 22:34:14 ##.##.50.11:45327 6:0-7:0> starting with 20 MB memory...
2017/08/24 22:34:14 ##.##.51.11:45327 2:0-3:0-4:0-5:0>2017/08/24 22:34:14 &{0xc42014e540}.CollectExecutionStatistics(_) = _, rpc error: code = Unavailable desc = grpc: the connection is unavailable

**Comment by chrislusf** [comment]
Need to also rebuild the binary for "gleam".

**Comment by carusyte** [comment]
I already did with the `go get -u github.com/chrislusf/gleam/...` command yesterday and I just double checked that the `distributed/executor/executor_grpc_client_to_agent.go` and `gio/runner_grpc_client_to_executor.go` were both updated to the latest version. The "gleam" binary executable was rebuilt, too, all before I ran the scenario.

**Comment by carusyte** [comment]
>2017/08/24 22:34:14 ##.##.50.11:45327 1:0>2017/08/24 22:34:14 executor heartbeat to :45327: executor collect stats from :45327 : rpc error: code = Unavailable desc = grpc: the connection is unavailable

I noticed that in the message, it says executor is heart beating to some address with 45327 port, but missing IP information, could that be suspicious? But I already started the agent by specifying the agent IP in "host" argument,
`gleam agent --master=##.##.51.10:45326 --host=##.##.50.11 --port=45327 --executor.max=8 --executor.cpu.level=2 --memory=8192`

**Comment by chrislusf** [comment]
it's trying to connect to localhost at port 45327.

**Comment by chrislusf** [comment]
added host name info to the connection. Please try it out.

This is only used to report execution stats though. It may not fix your empty result problem.

**Comment by carusyte** [comment]
When I deploy the word count example to a distributed environment, the output is lost, too. No idea what went wrong...

**Comment by chrislusf** [comment]
Please show the full command line and output of the master, agent, and
driver.

**Comment by carusyte** [comment]
The issue about word count and empty result in my driver program are both figured out now.

About word count, it's all because I changed the /etc/passwd file with another one in /etc/yum.repos.d, for security's sake, however in the assigned agent server this file does not exists. 

About my driver program, I made a stupid mistake by wrapping a nil error in a new one and return, and all the rest is a mess from then on. Sorry it wasted your time, too :)

Now, my new challenge is...
(is it idiomatic to post a new one instead of making this one endless?)

I create a flow as followed, all input data form as a 2d slice, partitioned and roundrobin-ed, still all the input is assigned to a single agent server. Don't know how it really works yet by reading the wiki and part of the examples and source code, perhaps digging deeper in the examples may help? But it would save me hours or even days if I could get some advice directly from you.
>	f := flow.New("KDJ Score Calculation").Slices(mapSource).Partition("partition",
		int(shard), sortOption).RoundRobin("rr", int(shard)).
		Map("kdjScorer", KdjScorer). // invoke the registered "kdjScorer" mapper function.
		ReduceBy("kdjScoreCollector", KdjScoreCollector, sortOption). // invoke the registered "kdjScoreCollector" reducer function.
		OutputRow(func(r *util.Row) error {
		for i, v := range r.V[0].([]interface{}) {
			rep.Scores[i] = v.(float64)
		}
		return nil
	})

**Comment by chrislusf** [comment]
you can skip Partition() function.

post your screenshot of the execution status url.

**Comment by carusyte** [comment]
<img width="1286" alt="screen shot 2017-08-28 at 9 17 11 am" src="https://user-images.githubusercontent.com/1680881/29756102-abccc4a2-8bd3-11e7-8e3e-57597f5aecf3.png">
-----------------------------------------------------------------------------------
<img width="1287" alt="screen shot 2017-08-28 at 9 18 02 am" src="https://user-images.githubusercontent.com/1680881/29756111-bc19e38a-8bd3-11e7-8d00-9b397edb4cfd.png">
-----------------------------------------------------------------------------------
<img width="1242" alt="screen shot 2017-08-28 at 9 18 39 am" src="https://user-images.githubusercontent.com/1680881/29756123-cd3fa67c-8bd3-11e7-9ed7-defe4dce7dad.png">
-----------------------------------------------------------------------------------
<img width="1230" alt="screen shot 2017-08-28 at 9 19 10 am" src="https://user-images.githubusercontent.com/1680881/29756176-512bf4ea-8bd4-11e7-8670-8cdebe4484dc.png">
-----------------------------------------------------------------------------------
<img width="1270" alt="screen shot 2017-08-28 at 9 19 35 am" src="https://user-images.githubusercontent.com/1680881/29756219-bdb7e4d4-8bd4-11e7-93d6-a87675a055c7.png">

**Comment by carusyte** [comment]
What I intended to do:
Divide a collection of data needing heavy calculation into independent "pieces" and dispatch each one of them to the agents. AFAIK, I transformed the collection into a [n][1] slice, submitted it as source of the flow. Struct element did not work so I made them maps. So each one of slice[n][0] is a complex map. I thought somehow each map would be dispatched to random agent for calculation, but I didn't make it.
